src/measure_theory/decomposition/signed_hahn.lean

```diff
@@ -0,0 +1,644 @@
+/-
+Copyright (c) 2021 Kexing Ying. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kexing Ying
+-/
+import measure_theory.vector_measure
+import order.symm_diff
+
+/-!
+# Unsigned Hahn decomposition
+
+This file defines positive and negative sets with respect to a vector measure and
+prove the Hahn decomposition (signed version).
+
+## Main definitions
+
+* `vector_measure.positive` : a set `i` is positive with respect to a vector measure if for all
+  measurable sets `j` contained in `i`, the measure of `j` is non-negative.
+* `vector_measure.negative` : a set `i` is negative with respect to a vector measure if for all
+  measurable sets `j` contained in `i`, the measure of `j` is non-positive.
+
+## Main results
+
+* `signed_measure.exists_disjoint_positive_negative_union_eq` : the Hahn decomposition theorem.
+* `signed_measure.exists_negative_set` : A measurable set of negative measure contains at least
+  one negative subset.
+
+## Tags
+
+positive, negative, Hahn decomposition
+-/
+
+noncomputable theory
+open_locale classical big_operators nnreal ennreal
+
+section lemmas
+-- ↓ I will move the content of this section to the approprate files after the reviews so we don't
+-- need to compile all of mathlib everytime.
+
+variables {α : Type*}
+
+open set filter
+
+lemma set.Union_inter_diff_eq {f : ℕ → set α} {a : set α} (ha : a ⊆ ⋃ n, f n) :
+  (⋃ n, a ∩ f n \ ⋃ k < n, f k) = a :=
+begin
+  ext x,
+  simp only [not_exists, exists_prop, mem_Union, mem_inter_eq, not_and, mem_diff],
+  split,
+  { rintro ⟨_, ⟨_, _⟩, _⟩, assumption },
+  { intro hx,
+    let n := nat.find (mem_Union.1 (ha hx)),
+    exact ⟨n, ⟨hx, nat.find_spec (mem_Union.1 (ha hx))⟩, λ m hm, nat.find_min _ hm⟩ }
+end
```

Comment:
I think it would be good to prove a more fundamental result first,
```lean
lemma set.Union_diff_Union_nat_lt (f : ℕ → set α) :
  (⋃ i, f i \ (⋃ k < i, f k)) = (⋃ n, f n) :=
eq_of_subset_of_subset
  (Union_subset_Union $ λ i, diff_subset _ _)
  (λ x hx, begin
    replace hx := mem_Union.1 hx,
    classical,
    simp only [mem_diff, mem_Union, not_exists],
    refine ⟨nat.find hx, nat.find_spec hx, λ m hm, nat.find_min hx hm⟩,
  end)
```
which makes this proof fall out as
```lean
lemma set.Union_inter_diff_eq {f : ℕ → set α} {a : set α} (ha : a ⊆ ⋃ n, f n) :
  (⋃ n, a ∩ f n \ ⋃ k < n, f k) = a :=
begin
  simp_rw [inter_diff_assoc, ←inter_Union, set.Union_diff_Union_nat_lt, inter_eq_left_iff_subset],
  exact ha,
end
```